CIP-???? | Ouroboros Leios - Greater Transaction Throughput (#1078)
cardano-foundation/CIPs: CIP-0164/README.md

### Protocol Flow

The protocol operates through five sequential steps that involve three critical
timing constraints. Figure 4 visualizes the precise timing mechanism that
governs when certificates can be safely included in the chain, showing both the
protocol parameters and the underlying network characteristics ($\Delta$ values)
and protocol parameters ($L$ values) that inform their design.

<div align="center">
  <a name="figure-4" id="figure-4"></a>
  <p name="protocol-flow-figure">
    <img src="images/protocol-flow.svg" alt="Leios Protocol Flow">
  </p>

<em>Figure 4: Detailed timing mechanism showing the three critical timing
constraints for EB certification</em>

</div>

#### Step 1: Block Production

Leios preserves the existing Praos chain structure while adding additional
transaction capacity through EBs. When a stake pool wins block leadership, it
may create two objects:

1. **[Ranking Block (RB)](#ranking-blocks-rbs)** A standard Praos block with
   extended header fields to optionally certify one previously announced EB and
   optionally announce one EB for the next subsequent RB (i.e., `RB'`) to
   optionally certify.

> 1. **[Endorser Block (EB)](#endorser-blocks-ebs)** A larger block containing

Reviewer: I realize we have not be giving a rationale for why we use transaction references and we should add it somewhere:

The protocol mainly discusses EBs for the purpose of endorsing a valid sequence of transactions. As transactions are diffused through the network already for the purpose of open participation, we avoid transmitting all of the endorsed transactions again by only including references to transactions. A consensus node would then only need to fetch the transactions it does not yet know about. This greatly reduces the amount of transmitted data and avoids re-transmission or re-validation when the same transaction gets endorsed by multiple EBs (e.g. when one EB did not get certified because of an "early" next RB).

For the purpose of storing the chain (immutably), we could imagine a more compact way of "inlining" transactions into EB bodies and chain validation reconstructing the hypothetical EB body (which was hashed and signed into the certificate), but we have not been going into details of storage in other places.

#### Step 3: Committee Validation

After the [equivocation detection period](#equivocation-detection) of
$3 L_\text{hdr}$ slots, a voting committee of stake pools validates the EB and
votes within a [voting period](#voting-period) $L_\text{vote}$. Committee
members are [selected via sortition](#committee-structure) based on the slot
number of the RB that announced the EB. A committee member votes for an EB only
if:

1. The RB header arrived within $L_\text{hdr}$,
2. It has **not** detected any equivocating RB header for the same slot,
3. It finished validating the EB before $3 \times L_\text{hdr} + L_\text{vote}$
   slots after the EB slot,
4. The EB is the one announced by the latest RB in the voter's current chain,
5. The EB's transactions form a **valid** extension of the RB that announced it,
6. For non-persistent voters, it is eligible to vote based on sortition using
   the announcing RB's slot number as the election identifier.

where $L_\text{hdr}$ and $L_\text{vote}$ are
<a href="#protocol-parameters">protocol parameters</a> represented by a number
of slots.

While not strictly a required check, honest nodes should not vote on empty EBs

> as that is obviously pointless and wasteful.

Reviewer 2: What does "not strictly" here imply? The formal spec **_strictly_** requires an honest node not vote for an empty EB: [`EndorserBlockOSig.txs eb ≢ []`](https://github.com/input-output-hk/ouroboros-leios-formal-spec/blob/c7f8d61865360cd16e04c7c3f0fd481c8081deff/formal-spec/Leios/Linear.lagda.md?plain=1#L135).

- Are we trying to say that a chain is valid even if it contains empty EBs but honest nodes should never vote for an empty EB?
- Or are we saying that we didn't need to include this required check in the specification but we did anyway for reasons of efficiency?

| <a id="l-diff" href="#l-diff"></a>Diffusion period length      | $L_\text{diff}$  |     slot     | Additional period after voting to ensure network-wide EB availability                       | Per [diffusion period](#diffusion-period): derived from the fundamental safety constraint. Leverages the network assumption that data known to >25% of nodes propagates fully within this time                                                   |
| Ranking block max size                                         |  $S_\text{RB}$   |    bytes     | Maximum size of a ranking block                                                             | Limits RB size to ensure timely diffusion                                                                                                                                                                                                        |

> | Endorser-block referenceable transaction size                  | $S_\text{EB-tx}$ |    bytes     | Maximum total size of transactions that can be referenced by an endorser block              | Limits total transaction payload to ensure timely diffusion within stage length                                                                                                                                                                  |

Reviewer 3: Is this to allow larger transactions to appear in endorser blocks than what's allowed by the current max Tx size parameter?

Reviewer: No. The `maxTxSize` would remain the same. The purpose of this new protocol parameter is to limit the total amount of validation / ledger state building that needs to be done when validating an EB. The EB size itself does not fully capture the "convex hull" of all bytes to validate. 

We should maybe say that explicitly in the [Size parameters](https://github.com/cardano-scaling/CIPs/blob/leios/CIP-0164/README.md#size-parameters) section.

| Maximum Plutus steps per transaction                           |        -         |  step units  | Maximum computational steps allowed for Plutus scripts in a single transaction within an EB | Limits computational resources per transaction to enable higher throughput                                                                                                                                                                       |
| Maximum Plutus memory per transaction                          |        -         | memory units | Maximum memory allowed for Plutus scripts in a single transaction within an EB              | Limits memory resources per transaction to enable higher throughput                                                                                                                                                                              |

<em>Table 3: Protocol Parameters</em>

</div>

#### Timing parameters

There are three key parameters related to time, which are important for

> [protocol security](#protocol-security). All relevant quantities are depdicted

Reviewer 3: ```suggestion
[protocol security](#protocol-security). All relevant quantities are depicted
```

  MsgLeiosBlockRangeRequest if Ouroboros Genesis is enabled.
- Recall that the `certified_eb` bit enables the client to correctly predict the
  total payload size of the valid replies to a MsgLeiosBlockRangeRequest. This
  enables the client to manage its receive buffers, balance load across peers,
  etc.
- A server should disconnect if the client requests an EB (or its transactions)
  the server does not have. For young EBs, a client can avoid this by waiting
  for MsgLeiosBlockOffer (or MsgLeiosBlockTxsOffer) before sending a request.
  For old EBs, a ChainSync MsgRollForward for an RB that certifies its
  predecessor's EB would also imply the server has selected that EB, and so must
  be able to serve it. Whether additional restrictions would be useful is not
  yet clear. For example, it seems natural to restrict MsgLeiosBlockRequest and
  MsgLeiosBlockTxsRequest to young EBs (and perhaps also
  MsgLeiosBlockRangeRequest to old EBs), but it's not already clear what the
  benefit would be.
- If MsgLeiosBlockRequest and MsgLeiosBlockTxsRequest were restricted to young
  EBs, then MsgLeiosBlockRangeRequest would not only enable syncing nodes but

>   also the unfortunate node that suffers from a $\Delta^\text{A}_\text{EB}$

Reviewer 4: This Delta^A might be a typo (Delta^W perhaps?), I do not see it defined elsewhere in the document.

$$

**_Network efficiency:_** Effective utilization of the network can be
characterized by the ratio of bytes of transactions reaching the ledger to the
average network traffic per node. (This could also be computed individually for
each node and used as a local metric.)

$$
\epsilon_\text{network} = \frac{(\text{bytes of valid transactions reaching the ledger}) \cdot (\text{number of nodes in the network})}{\text{total bytes of network traffic}}
$$

**_TX inclusion:_** In Leios, it is possible that a transaction might have to
wait for multiple EB production opportunities before being included in an EB.
The characteristic time for such inclusion in an EB depends on the EB production
rate and mempool management. This is correlated with how long the transaction
waits in the memory pool before being selected for inclusion.

$$
\tau_\text{inclusion} = \text{mean number of slots for a transaction to be included in any EB}
$$

**_Voting failure:_** An unlucky set of VRF evaluations might result in
insufficient voters being selected in a given pipeline, thus making it

> impossible to certify an EB in that pipeline.

Reviewer 4: Does the *pipeline* terminology correspond to linear Leios, or is a remnant from the Research/Full version?

  updating the current ledger state:
  - CPU per transaction in a block: `620.1 μs/tx`.
  - Linear model that accounts for Plutus:
    `(262.4 μs/tx) * (number of transactions) + (948.7 μs/Gstep) * (billions of Plutus execution steps)`.
- Ledger "reapply" operation, consisting of updating the current ledger state,
  omitting previously performed phase 1 & 2 validation:
  - Linear model: `(353.9 μs) + (21.51 μs/kB) * (size of the block)`
  - Linear model that accounts for Plutus:
    `(347.8 μs) + (19.43 μs/kB) * (size of the block) + (21.27 μs/Gstep) * (billions of Plutus execution steps)`

The Leios simulators perform the "apply" operation when a transaction is first
seen, either when it is received for the memory pool or when it is fetched after
first being seen in an RB or EB; they perform the "reapply" operation when a
block is being generated or validated. A more nuanced model of CPU usage in the
simulators would account for Plutus execution explicitly, but the linear models
described above are used to account for Plutus workloads implicitly. The
following plot of simulation results limit each node to 4 vCPU cores and suggest

> that workloads of 20,000e9 Plutus execution steps per EB may be feasible: this
> is 1000 times the current Cardano mainnet limit of 20e9 steps for Praos blocks.

Reviewer 4: ```suggestion
that workloads of 2e13 Plutus execution steps per EB may be feasible: this
is 1000 times the current Cardano mainnet limit of 2e10 steps for Praos blocks.
```
Perhaps use the `e` notation like this, or something more explicit like $20\cdot 10^{12}$.

</div>

<div align="center">
<a name="figure-18" id="figure-18"></a>

|                                                                                             |                                                                                             |
| ------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------- |
| ![Mean CPU usage in Plutus-heavy workloads for Leios](images/plutus-cpu-mean-histogram.svg) | ![Peak CPU usage in Plutus-heavy workloads for Leios](images/plutus-cpu-peak-histogram.svg) |

<em>Figure 18: CPU usage in Plutus-heavy workloads for Leios</em>

</div>

In summary, Leios will require a modest increase of the [recommended hardware
requirements][spohw]: a four-core machine will be required, but a network
upgrade will not be needed, as 10 Mb/s is well below the bandwidth of standard
network connections. At throughput much higher than 200 kB/s, network egress can
become a significant cost for nodes hosted on some cloud-computing providers.
The Leios simulations do not model memory or disk. With the advent of

> [UTxO-HD][utxohd], 16 GB of memory will remain be sufficient for Leios if the

Reviewer 4: Perhaps the link https://ouroboros-consensus.cardano.intersectmbo.org/docs/references/miscellaneous/utxo-hd/ is more appropriate here

`OnDisk` option is used for the UTxO set. Disk requirements depend upon the
growth of the ledger, but a sustained 0.150 MB/s throughput amounts to ledger
size increasing by 4.7 TB each year: see the section below on Operating Costs
for further discussion.

### Feasible Protocol Parameters

**Parameter Relationships and Network Assumptions**

The key relation in the proposed protocol is between the voting threshold

> ($\tau = 75\%$) and propagation delay of EBs ($\Delta_\text{EB}$). The high

Reviewer 4: The % sign from tau does not render in my Debian desktop (Test here, `$100\%$`: $100\%$). Do you see it on your end?

recent Praos limit on transaction execution steps. The per-transaction limit can
be adjusted to suit the needs of the community: i.e., it could be tuned to favor
many light Plutus transactions vs a few heavy Plutus transactions.

Although the Praos maximum block size could be modestly raised in Leios and the
active-slot coefficient adjusted slightly, there is no compelling reason to
alter these. They could, however, be re-evaluated in the context of the Leios
testnet.

The analysis [Committee size and quorum requirement][committee-size-analysis] in
the first Leios Technical Report indicates that the Leios committee size should
be no smaller than 500 votes and the quorum should be at least 60% of those
votes. However, the proposed [Fait Accompli][fait-accompli-sortition] scheme
wFA<sup>LS</sup> achieves compact certificates that do not become larger as the
number of voters increases, so larger committee sizes might be permitted for
broader SPO participation and higher security. The committee size should be
large enough that fluctuations in committee membership do not create an
appreciable probability of an adversarial quorum when the adversarial stake is

> just under 50%. The quorum size should be kept large enough above 50% so that

Reviewer 4: ```suggestion
just under 50%. The quorum size should be kept just large enough above 50% so that
```
Is this the intended meaning?

| -------------------------------------------------------------------- | :---------------------------: | --------------------------------------------------------------------------------------------------------------------------- | :---------------------------: |
| <a id="delta-hdr" href="#delta-hdr"></a>Header propagation           |      $\Delta_\text{hdr}$      | Time for constant size headers (< 1,500 bytes) to propagate network-wide                                                    |            <1 slot            |
| <a id="delta-rb" href="#delta-rb"></a>RB diffusion                   |      $\Delta_\text{RB}$       | Complete ranking block propagation and adoption time                                                                        |           <5 slots            |
| <a id="delta-eb-H" href="#delta-eb-H"></a>EB optimistic diffusion    | $\Delta_\text{EB}^{\text{O}}$ | EB **diffusion** time (transmission + processing) under favorable network conditions                                        |           1-3 slots           |

> | <a id="delta-eb-A" href="#delta-eb-A"></a>EB worst-case transmission | $\Delta_\text{EB}^{\text{W}}$ | EB **transmission** time for certified EBs starting from >25% network coverage (processing already completed during voting) |          15-20 slots          |

Author: Thanks for pointing this out. I've updated the time intervals from logical to physical.

$\Delta_\text{RB}$ slots, due to the fact that it is received within
$\Delta_\text{RB} - \Delta_\text{applyTxs}$ from its initial diffusion time, and
processing in the worst-case takes
$\Delta_\text{reapply} (< \Delta_\text{applyTxs})$ slots according to
[Constraint 1](#eb-reapplication-constraint).

Given that nodes are caught up when they are about to produce or process an RB,
Praos safety and liveness is thus preserved.

### Protocol Artifacts

The protocol extends Praos blocks, introduces endorser blocks, Leios votes and
certificates. While already introduced briefly, this section specifies these
artifacts in more detail.

#### Ranking Blocks (RBs)

RBs are Praos blocks extended to support Leios by optionally announcing EBs in
their headers and embedding EB certificates in their bodies.

1. **Header additions**:
   - `announced_eb` (optional): Hash of the EB created by this block producer
   - `announced_eb_size` (optional): Size in bytes of the announced EB (4 bytes)
   - `certified_eb` (optional): Single bit indicating whether this RB certifies
     the EB announced by the previous RB (the EB hash is already available via

>      the previous header's `announced_eb` field)

Author: Thanks, this is generally a starting point for the CBOR and we understand that the engineering team collectively will settle on a better form which we will update the CIP for. The specific CDDL is defined in the appendix. Can you propose concretely an improved version?

>      the previous header's `announced_eb` field)

Reviewer 5: There are already block type IDs that correspond to the various block types within an era (there are 2 for Byron and currently only 1 for each subsequent era). The EB and non-EB variations of the blocks and their headers could be uniquely identified by giving them separate block type IDs.

include the `endorser_block_hash` field that uniquely identifies the target EB:

- **Persistent votes**:
  - `election_id`: Identifier for the voting round (derived from the slot number
    of the RB that announced the target EB)
  - `persistent_voter_id`: Epoch-specific pool identifier
  - `endorser_block_hash`: Hash of the RB header that announced the target EB
  - `vote_signature`: Cryptographic signature (BLS in this implementation)
- **Non-persistent votes**:
  - `election_id`: Identifier for the voting round (derived from the slot number
    of the RB that announced the target EB)
  - `pool_id`: Pool identifier
  - `eligibility_signature`: Cryptographic proof of sortition eligibility (BLS
    in this implementation)
  - `endorser_block_hash`: Hash of the RB header that announced the target EB
  - `vote_signature`: Cryptographic signature (BLS in this implementation)

The `endorser_block_hash` identifies the header that announces the EB instead of
identifying the EB's hash directly. This ensures the voters validated the EB
against the same ledger state that it extends when certified on chain; recall

> that multiple RB headers could announce the same EB.

Reviewer 5: This name doesn't seem to match with what it represents. Perhaps it should be called something like `endorser_block_announced_rb_hash`

Author: Thanks, I've refactored the name.

timing constraints that ensure RB processing stays within Praos bounds. The
high-participation design (75% voting threshold) eliminates invalid transactions
in RBs and provides strong network assumptions for certified EB propagation. New
threats include equivocation by EB producers/voters and transaction availability
attacks, mitigated through cryptographic validation, equivocation detection, and
the high voting threshold requirement. Comprehensive analysis is documented in
the [Protocol Security](#protocol-security) section and
[threat model](https://github.com/input-output-hk/ouroboros-leios/blob/d5f1a9bc940e69f406c3e25c0d7d9aa58cf701f8/docs/threat-model.md).

### Design Decisions

#### Transaction References in Endorser Blocks

EBs contain transaction references rather than full transaction bodies to avoid
re-transmitting data already diffused through the network. Nodes fetch only
transactions they don't possess, reducing bandwidth and avoiding redundant
validation when transactions are endorsed by multiple EBs. Complete transaction
data remains available through network diffusion and local storage for immutable

> chain purposes.

Reviewer 5: This seems to bury the lede that EB TXs must be stored out-of-band from RBs and EBs, since the TX body will not be contained in either.

| ←Server | MsgLeiosVoteDelivery            | list of votes                                                | The votes from an earlier MsgLeiosVotesRequest.                                                                                                                                                                                       |
| Client→ | MsgLeiosBlockRangeRequest       | two slots and two RB header hashes                           | The server must now deliver the EBs certified by the given range of RBs, in order.                                                                                                                                                    |
| ←Server | MsgLeiosNextBlockAndTxsInRange  | an EB and all of its transactions                            | The next certified block from an earlier MsgLeiosBlockRangeRequest.                                                                                                                                                                   |

> | ←Server | MsgLeiosLastBlockAndTxsInRange  | an EB and all of its transactions                            | The last certified block from an earlier MsgLeiosBlockRangeRequest.                                                                                                                                                                   |

Reviewer 5: Is every single one of these messages required in a typical exchange? This table seems to imply that a server must send an EB announcement, a "block available" announcement, a "block and TXs available" announcement, and a votes announcement before the client can start fetching any of those things. Should a client treat everything short of "block and TXs available" and "votes available" as advisory rather than something to act on?

This appendix contains the complete CDDL specifications for all Leios protocol
messages and data structures. These definitions specify the exact wire format
for network communication.

<a id="ranking-block-cddl" href="#ranking-block-cddl">**Ranking Block**</a>

```diff
 ranking_block =
   [ header                   : block_header
   , transaction_bodies       : [* transaction_body]
   , transaction_witness_sets : [* transaction_witness_set]
   , auxiliary_data_set       : {* transaction_index => auxiliary_data}
+  , ? eb_certificate         : leios_certificate
   ]

block_header =
   [ header_body              : block_header_body
   , body_signature           : kes_signature
   ]

 block_header_body =
   [ block_number             : uint
   , slot                     : slot_no
   , prev_hash                : hash32
   , issuer_vkey              : vkey
   , vrf_vkey                 : vrf_vkey
   , vrf_result               : vrf_cert
   , block_body_size          : uint
   , block_body_hash          : hash32
+  , ? announced_eb           : hash32
+  , ? announced_eb_size      : uint32
+  , ? certified_eb           : bool

>    ]

Reviewer: `announced_eb` and `announced_eb_size` may only be used together. So making them a structure would make sense to me.

The `certified_eb` is independently optional though.

>    ]

Author: Thanks @agaffney I've updated the CDDLs respectively.

not yet received MsgLeiosBlockTxsOffer.

<div align="center">
<a name="table-4" id="table-4"></a>

| Sender  | Name                            | Arguments                                                    | Semantics                                                                                                                                                                                                                             |
| ------- | ------------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Client→ | MsgLeiosNotificationRequestNext | $\emptyset$                                                  | Requests one Leios notifications, the announcement of an EB or delivery offers for blocks, transactions, and votes.                                                                                                                   |

> | ←Server | MsgLeiosBlockAnnouncement       | RB header that announces an EB                               | The server has seen this EB announcement.                                                                                                                                                                                             |

Reviewer 6: @dcoutts, thanks for the questions you raised in this comment. https://github.com/cardano-foundation/CIPs/pull/1078#issuecomment-3356670304

I'm opening this comment localized on this "definition" of `MsgLeiosBlockAnnouncement` because I think it's particularly important one of the main concerns in your questions. And also so that we can have a threaded convesration.

Reviewer 6: We discussed some of your questions during the call earlier today, so I'll be a little brief in the responses. Please feel free to (re-)object to parts of my explanation here that seem dissatisfying.

> I assume that we're also separating the notion of a node's current chain from the notion of what prefix of the node's chain has a ledger computed for it. With the idea that we can quickly validate and select an alternative fork, without having to compute the ledger for that fork.

> The bit that seems problematic to me is the idea that we receive all EBs from competing forks. At a minimum, it's unclear what this really means and how that could be implemented given the existing Praos protocols (which apparently are not intended to be changed).

> And how with the existing protocols? A node only sees one upstream chain for each peer at once. So "all" forks are not visible to all nodes. Worse forks are inherently pruned in the relaying process by selecting only the best chain at each hop.

The key point that this question brings to mind for me is that MsgLeiosBlockAnnouncement sends the RB header independently of ChainSync and moreover independently of chains. Only opcert issue numbers are influenced by recent headers, so the client receiving MsgLeiosBlockAnnouncement doesn't even need to have received the preceding header. (There're more details written about the consequences of the missing opcert issue numbers elsewhere.)

> What does it really mean? Any valid fork (within the last K) blocks? For example, if I (as an adversary) make 1-block extensions of old nodes, do I expect every other node in the network to go back and fetch an EB for that fork just in case this chain could become a longest/densest chain? Surely not. That sounds like a DoS attack.

The intended idea is that the node will have acquired _every_ EB as they arise, even if those EBs are not on the node's current chain. And the conservative rule is for it would retain those EB until the node's immutable tap implies it could never select a chain that includes that EB.

A key factor is that the Leios anti-equivocation mechanisms prevent the node from ever downloading more than one EB per Praos election (or maybe it's 2). I don't remember the exact math, but the ballpark number from the last time Esgen walked me through the appropriate probability bounds was that even with grinding there couldn't be more than ~11,000 elections per 36 hours on `mainnet`. 12.5 megabytes * 2 * 11,000 is 275 gigabytes, which is frighteningly large but also isn't completely intractable worst-case. And a less conservative rule (eg 6 hrs/not assuming grinding/etc) might be justifiable.

> And in general it's not possible to see alternative forks early, and thus not possible to see EBs early. If we have a temporary 2-way network split, then by definition, nodes in each partition cannot see the chain(s) from the other partition. If those partitions are re-connected within K blocks then at that moment, the chains become visible, but the EBs will necessarily have to be fetched at this point. They cannot be fetched earlier.

Partitions do pose a challenge. If they're very lopsided, then the big side could be producing certs and the small side would be forced to fetch the EBs once the partition ends and could not contribute until it finishes catching-up. If the partition is more so balanced such that no side is "big", then there are no certs, and so there will never be a need for either side to recover the EBs they missed out on when the partition ends. The fact that the adversary can be fully present in multiple sides of a partition does complicate those claims a bit, but with a cert threshold of 75%, the adversary would have to have 50% stake in order to cause multiple sides of a perfect-storm partition to create certs.

The optimistic perspective can be summarized by noting that only the small side(s) of a partition would be delayed recovering EBs that the big side of a partition was able to certify during the partition. The small side grows its chain more slowly since it's the small side, so the partition would have to last for quite some time before that side risked having it's own 2161st block. Even the delay of having to fetch however many EBs is likely a small fraction of the duration of a partition that would introduce the risk of the small side growing its own 2161st block (unless the adversary was also issuing blocks on that side's chain with its full, say, 40% stake).

---

We should criticize that argument and we should explicitly write something in the CIP about partitions.

from the `mini-mainnet` topology are also valid for the `pseudo-mainnet`
topology; the advantage of using the former is that simulations run much more
quickly. Simulated RB diffusion is consistent with the Praos performance
model.[^praosp]

[^mnrm]:
    [Mainnet-like topologies for Leios](https://github.com/input-output-hk/ouroboros-leios/blob/6d8619c53cc619a25b52eac184e7f1ff3c31b597/data/simulation/pseudo-mainnet/ReadMe.md)

[^pseudo]:
    [Leios pseudo-mainnet topology](https://github.com/input-output-hk/ouroboros-leios/blob/6d8619c53cc619a25b52eac184e7f1ff3c31b597/data/simulation/pseudo-mainnet/topology-v1.md)

[^mini]:
    [Leios mini-mainnet topology](https://github.com/input-output-hk/ouroboros-leios/blob/6d8619c53cc619a25b52eac184e7f1ff3c31b597/data/simulation/pseudo-mainnet/topology-v2.md)

[^ripe]: [RIPE Atlas](https://atlas.ripe.net/)

[^mncp]:
    https://github.com/input-output-hk/ouroboros-leios/blob/6d8619c53cc619a25b52eac184e7f1ff3c31b597/analysis/sims/2025w30b/analysis.ipynb

[^praosp]:
    https://github.com/IntersectMBO/cardano-formal-specifications/blob/6d4e5cfc224a24972162e39a6017c273cea45321/src/performance/README.md

> The simulation results in the remainder of this section use the Rust simulator

Author: As far as I know, the current rust simulation does **not** have a representation of UTxOs as such with the only exemption of a single fee output for simulating a sharded ledger design (which isn't required for linear leios). 
Though your data suggests that an implementation might actually benefit from some caching strategy for recent newly created outputs being spent within a certain time.

I'll reference this data for the implementation team to consider. Thanks for sharing again and please let me know if your questions are resolved.

> The simulation results in the remainder of this section use the Rust simulator

Reviewer 4: Yes, this is resolved, thank you!

<a name="figure-16" id="figure-16"></a>

![Mean CPU load among all nodes](images/cpu-mean-histogram.svg)

<em>Figure 16: Mean CPU load among all nodes ("Gen" = generated, "Val" =
validated, "RH" = ranking block header, "RB" = ranking block body, "EH" =
endorser block header, "EB" = endorser block body", "TX" = transaction)</em>

</div>

Note that the transaction workload in the simulations above was modeled upon the
_average_ amount of Plutus computation typical of the Cardano mainnet. The low
time-averaged CPU usage in the simulations (i.e., less than 15% of a vCPU)
suggests that the per-transaction and/or per-block Plutus budget could be
significantly increased under Leios: either every transaction could have a
modestly higher budget, or some transactions could use an order of magnitude
more Plutus execution units. Statistical analysis of [CPU usage in ledger
operations][timings] using the [db-analyser tool][dbanalyser] on Cardano mainnet
from epoch 350 through 573 yields the following simple models of the CPU cost of
validating signatures and executing Plutus in the transactions of a block.
Because of the noisiness in the raw mainnet data, these estimates are uncertain.

> - Ledger "apply" operation, consisting of phase 1 & 2 validation along with

Author: Thanks for sharing this data @sterraf and yes it's correct our simulations can be configured how transaction sizes are distributed. (see [here](https://github.com/input-output-hk/ouroboros-leios/blob/main/data/simulation/config.default.yaml#L56-L59)).

I'll see if we can use this data to improve our simulations. Please let me know if you have other questions or if I can resolve this.

> - Ledger "apply" operation, consisting of phase 1 & 2 validation along with

Reviewer 7: @sterraf what do you mean by Plutus budgets? We don't simulate Plutus directly; a transaction's size is controlled by a distribution, and its CPU cost is [controlled by its size](https://github.com/input-output-hk/ouroboros-leios/blob/main/sim-rs/implementations/LINEAR_LEIOS.md#cpu-model).

   based on Praos KES signatures are too large for consideration in Leios.

9. **Fast cryptography:** The computational burden of creating and verifying
   voting eligibility, the votes themselves, and the resulting certificate must
   be small enough to fit within the CPU budget for Leios stages.

The aggregate signature scheme implementation specified in this document (using
BLS as an example) satisfies all these requirements, as evidenced by the
performance characteristics and certificate sizes documented in the
[Votes and Certificates](#votes-and-certificates) section.

<h3 id="appendix-b-cddl">Appendix B: Wire Format Specifications (CDDL)</h2>

This appendix contains the complete CDDL specifications for all Leios protocol
messages and data structures. These definitions specify the exact wire format
for network communication.

<a id="ranking-block-cddl" href="#ranking-block-cddl">**Ranking Block**</a>

```diff
 ranking_block =
   [ header                   : block_header
   , transaction_bodies       : [* transaction_body]
   , transaction_witness_sets : [* transaction_witness_set]
   , auxiliary_data_set       : {* transaction_index => auxiliary_data}

> +  , ? eb_certificate         : leios_certificate

Reviewer 4: I'm giving it a try at implementing Leios primitives at Pallas.
I see that `invalid_transactions` is missing here; did you think of keeping that field in the last position, or before `eb_certificate`?

<a id="ranking-block-cddl" href="#ranking-block-cddl">**Ranking Block**</a>

```diff
 ranking_block =
   [ header                   : block_header
   , transaction_bodies       : [* transaction_body]
   , transaction_witness_sets : [* transaction_witness_set]
   , auxiliary_data_set       : {* transaction_index => auxiliary_data}
   , invalid_transactions     : [* transaction_index]
+  , ? eb_certificate         : leios_certificate
   ]

block_header =
   [ header_body              : block_header_body
   , body_signature           : kes_signature
   ]

 block_header_body =
   [ block_number             : uint
   , slot                     : slot_no
   , prev_hash                : hash32
   , issuer_vkey              : vkey
   , vrf_vkey                 : vrf_vkey
   , vrf_result               : vrf_cert
   , block_body_size          : uint
   , block_body_hash          : hash32
+  , ? ( announced_eb         : hash32
+      , announced_eb_size    : uint32
+      )
+  , ? certified_eb           : bool
   ]
```

<a id="endorser-block-cddl" href="#endorser-block-cddl">**Endorser Block**</a>

```cddl
endorser_block =
  [ transaction_references   : [* tx_reference]

>   ]

Reviewer 8: You are both correct! :smile: 

Having just sequence of transactions is incorrect, since then you could have different multiple entries in the list with duplicate hashes but different size. In such a case when it would be decoded in the implementation, with that implementation using a Map, then one of those entries would disappear. That would be bad, despite that by definition one of those entries would be invalid (since changing the size will change the hash). In other words it would allow for EB to have invalid entries in the serialization which would get ignored, which one exactly would get ignored would be implementation specific. We had occurrences of such incorrect definition in Ledger, which we have fixes since.

On the other hand, despite that using a map solves the problem with duplicates, it does not preserve order, which is also important.

The proper solution is to use an ordered map here. CDDL does not have an official definition like that, but we do already have in Ledger CDDL examples of ordered sets. In a similar fashion we can add definition of an ordered map. We do have support for such data structure in Haskell already and use it for some data in the ledger state.

In other words, I would suggest accepting the suggestion with added comment that this Map definition will use order preserving (de)serialization. From what I know CDDL does not say anything about the order or how duplicates are handled, so we can choose how we interpret this map definition.

<div align="center">
  <a name="figure-3" id="figure-3"></a>
  <p name="protocol-overview">
    <img src="images/protocol-overview.svg" alt="Leios Chain Structure">
  </p>

<em>Figure 3: Leios chain structure showing the relationship between Ranking
Blocks, Endorser Blocks, and Certificates</em>

</div>

The horizontal spacing in Figure 3 reflects the opportunistic nature of EB
inclusion: some EBs get certified and are included in the chain (green), while
others cannot be certified in time (gray). This happens because Praos block
production is probabilistic - some RBs will naturally occur before there has
been sufficient time for the preceding EB to gather the necessary votes and
certification, invalidating these EBs to become adopted. The key insight is that
proposed Leios utilizes the natural intervals between Praos blocks to perform
additional work on transaction processing without interfering with the base
chain operation.

EB inclusion is therefore opportunistic rather than guaranteed, depending on the
random timing of block production relative to the certification process. The
precise timing mechanism is detailed in the following section.

> ### Protocol Flow

Reviewer: @fallen-icarus I do agree with your assessment of MEV opportunities in DeFi and thanks for the example link!

> With Full Leios, this instant gratification is not possible. If a malicious SPO sees an EB with a profitable trade, that SPO can't simply add their own version of that tx into the RB. Instead, they must add it to an IB and send it through the entire Leios pipeline for a future SPO to possibly place in an RB. This is because in Full Leios, RBs can only hold EBs, not individual txs.

Which protocol specification is this referring to? IIRC the published paper never went into this detail, neither did the R&D material I came across. To my understanding any of the Leios variants must allow RBs to directly hold txs or it would endanger safety (and by extension liveness) properties, for a certificate may never be produced given enough adversarial stake. So unfortunately your conclusion of "why this is a non-issue in Full Leios" doesn't hold.

(The next paragraphs are not specifically directed to anyone)

A possible mitigation in decoupled block production variants - like the published ones or shorter pipeline forms - is, _if_ we'd have the ledger order transactions (or IBs) by their slot instead. This would lead to possible re-ordering of transactions on the same chain when an EB is included; i.e. at the cost of introducing another dimension of non-monotonicity to the ledger (besides forks). Note, that this would also invalidate the transactions included earlier by a later EB and make them "not paid for" because their fees would not be collected (unless we deem them also as "failed"?).

Does it make sense to compare the proposed design with less detailed, alternative designs? IMO we should keep focus on how this proposal would improve Cardano (or not) and compare it with what is currently deployed.

Two points come to mind:
- To protect a tx from being front-run in Cardano today, one must not forward such a tx to other block producers and include it directly in a block (or trust an SPO to that for you). The same front-running protection is available with the Linear Leios protocol. It will harm throughput, but the remaining throughput will be strictly higher than with Praos alone. Besides, the increased volume of activity to front run does not exist and can't exist with Praos alone because there is no demand without capacity (the premise of this CIP as identified by CPS-18).
- More a silver lining: With Linear Leios, front-running is easier to detect than today. That is, if a transaction is part of an EB, which gets certified and L_diff has passed, but an SPO does _not_ include the cert and instead chooses to forge a block that conflicts with the certified tx set, it's a strong indicator they deliberately did so. Still plausible deniable though, unless they also voted for the EB. I'm not implying an incentive scheme here, but just want to mention _detect-ability_.

@pagio May I invite your commentary here as we had discussed front-running a few times over the last weeks?

> ### Protocol Flow

Reviewer 9: Thank you guys for the comments. My two cents on this one:

> Right now, an SPO can only see what is inside their mempool - about 200 kb. …  Odds of finding a profitable MEV opportunity sky-rocket and the instant gratification possible with Linear Leios make it possible to take advantage of any opportunities found in this extra 10 mb of txs.

*TLDR*: Linear Leios maintains the same or a lower level of risk exposure of users to MEV attacks.

A rational/MEV-hunting SPO in Praos can easily observe a higher amount of traffic than 200 kb, e.g., by increasing the mempool limit, setting up multiple nodes in different parts of the network, etc. Hence, publicly available MEV opportunities will be known to rational nodes early on, and not after they appear in an RB. 

Further, these txs will most likely be exposed to a higher risk of being front-run in Praos. Given the limited throughput of Praos compared to Linear Leios, it will take longer for a tx of this type to be included in a block, and thus more block producers are going to consider it for inclusion. Hence, it is more likely that the tx is picked up by a rational block producer,that will try to extract value from it.

Finally, users can avoid publicizing their tx and getting exposed to MEV attacks in Linear Leios in the same way as in Praos: by coordinating with specific SPOs and having their txs being directly included in an RB. Note, that this service is naturally supported in Linear Leios, as for the parameters considered in the CIP, by design for a large fraction of RBs there will be no certified EB to be included due to the last RB being generated within the previous RB’s diffusion period, and instead the RB will directly contain txs.

> ### Protocol Flow

Reviewer 10: @pagio @berewt I think my comments are being misunderstood. **I am not trying to eliminate MEV** - that's not possible. **I am also not suggesting a change to Linear Leios**. I am trying to point out **a potential trust assumption** we can use instead of enforcing anything at the protocol level. Instead of evaluating this trust assumption, the discussion keeps going on tangents about MEV...

The core problem is that txs in an RB means there is no throughput boost for that block. As a crisis fallback or in the case of low demand, this is fine. But if txs are placed in RBs under non-crisis/high-demand periods, this is not good! The MEV prevention service that keeps being suggested is not good for this reason. If MEV happens inside an EB, that is fine because we still get the throughput boost for that block. But if MEV results in txs in RBs, that forces us to sacrifice significant throughput.

So the question I've been trying to point out is: _how do we prevent txs in RBs even in the presence of MEV?_ I think we can solve this with just a social contract. Below is the second half of my prior [comment](https://github.com/cardano-foundation/CIPs/pull/1078#discussion_r2549920094) that seems to have been missed. **It does not need to be baked into Linear Leios!** If people just delegate away from SPOs who deviate from the norm, that is good enough.

> 1. For all stake pools, track their ratio of RBs with txs vs EB certificates over a given time-frame. Use this to generate a global average ratio.
> 2. If an SPO's individual average is ever meaningfully higher than this global average, either it has a technical issue or it is frequently using the RB to front-run. Either way, honest delegators should delegate away.
>
> If there is actually a systemic problem [or low systemic demand], the global average of the tx/EB ratio should rise. So as long as the majority of honest stake prefers throughput over [txs] in RBs, the global average should be near optimal. In this context, an SPO trying to [put txs in RBs] too frequently will cast an immediate spotlight on their bad behavior and they will lose delegators. 

Under this pressure, MEV can still happen inside EBs but that isn't as bad because it doesn't sacrifice significant throughput. As long as the community enforces this global ratio, the MEV prevention service that keeps being suggested can work similarly by using EBs instead of txs in RBs.

Again, **this doesn't require a change to Linear Leios**. It is a social contract, not a protocol enforced metric.

> ### Protocol Flow

Reviewer 9: > If the attacker signs the cert and sends it to even a single honest SPO, wouldn't it just get propagated to everyone else automatically? 

@fallen-icarus  Propagation is not automatic. It takes some time for certificate to diffuse around the network. This is the window an attacker would try to take advantage off, by sending his votes (and thus the certificate in some sense) at the right time to say half the network see it as on time and the other half see it as delayed.

> doesn't Cardano's private leader schedule prevent the attack you mention? 

The attack is based on the attacker observing the block and then reacting (i.e., voting). So the private schedule does not help much in this case.

> ### Protocol Flow

Reviewer 10: @pagio The attack you describe would not work in the system I am imagining. Consider an entirely off-chain **Mithril-based Availability Monitoring** system. Whenever an RB contains txs instead of an EB cert, a poll can go out to all SPOs: "Did you see an EB cert that met the `X%` stake threshold before it was time to produce that RB?" To declare the RB producer's action as malicious, the poll would need to achieve the *same stake threshold* as the EB certificate itself. So your scenario would fail because the poll would not meet the required threshold of "Yes".

Some important points about the monitoring system (as I see it):
- **It is purely off-chain.** It has no impact on consensus and requires no changes to the Linear Leios protocol.
- **It is for observation only.** It is just a community tool for logging and transparency.
- **Data can be pruned regularly**. The results are only needed for short-term accountability. We don't need it to replay the chain.

This mithril-based polling system may also be useful for observing other parts of the distributed system.

---
CIP: 164
Title: Ouroboros Linear Leios - Greater transaction throughput
Status: Proposed
Category: Consensus
Authors:
  - William Wolff <william.wolff@iohk.io>
  - Brian W Bush <brian.bush@iohk.io>
  - Sebastian Nagel <sebastian.nagel@iohk.io>
  - Nicolas Frisby <nick.frisby@iohk.io>
  - Giorgos Panagiotakos <giorgos.panagiotakos@iohk.io>
  - Andre Knipsel <andre.knispel@iohk.io>
  - Yves Hauser <yves.hauser@iohk.io>
  - Simon Gellis <simon@sundae.fi>
Implementors:

>   - Intersect

Reviewer 11: I feel this is probably more accurate

Certificates**</a>

```cddl
leios_certificate =
  [ election_id              : election_id
  , endorser_block_hash      : hash32
  , persistent_voters        : [* persistent_voter_id]
  , nonpersistent_voters     : {* pool_id => leios_bls_signature}
  , aggregated_vote_sig      : leios_bls_signature
  ]

leios_vote = persistent_vote / non_persistent_vote

persistent_vote =
  [ 0
  , election_id
  , persistent_voter_id
  , endorser_block_hash
  , vote_signature : leios_bls_signature
  ]

non_persistent_vote =
  [ 1
  , election_id
  , pool_id
  , eligibility_signature : leios_bls_signature
  , endorser_block_hash
  , vote_signature : leios_bls_signature
  ]
```

Leios uses **BLS12-381 MinSig** (small signature, large verification key):

- Verification key: compressed G2 = 96 bytes
- Signature: compressed G1 = 48 bytes
- Proof-of-possession: compressed G1 = 48 bytes

MinSig is chosen because:

- Certificate size is ~8 kB vs >12 kB for MinPk
- Network propagation and storage are significantly improved

```cddl
; BLS12-381 MinSig encodings for Leios
leios_bls_verification_key = bytes .size 96
leios_bls_signature        = bytes .size 48

> leios_bls_pop              = bytes .size 48

Reviewer: I assume this popped up while cross checking the implementation in `cardano-base` https://github.com/IntersectMBO/cardano-base/pull/587 here? I just thought I leave the link here for reference.
